contents/tutorials/python-ab-testing.md

```diff
@@ -0,0 +1,181 @@
+---
+title: How to set up Python A/B testing
+date: 2023-09-07
+author: ["ian-vanagas"]
+showTitle: true
+sidebar: Docs
+featuredImage: ../images/tutorials/banners/tutorial-14.png
+tags: []
+---
+
+A/B testing enables you to experiment with how changes to your app affect metrics you care about. PostHog makes it easy to set up [A/B testing](/ab-testing) in Python. This tutorial shows you how to create a basic Python app with Flask, add PostHog to it, and then set up an A/B test to compare button variants.
+
+## Creating a basic Flask app
+
+To showcase A/B testing, create an app using [Flask](https://flask.palletsprojects.com/), a Python web framework. To start, create a folder for our app named `ab-test-demo` and a file named `hello.py` in that folder.
+
+```bash
+mkdir ab-test-demo
+cd ab-test-demo
+touch hello.py
+```
+
+Next, create a virtual environment named `venv` for our app, activate the virtual environment, and install Flask.
+
+```bash
+python3 -m venv venv
+. venv/bin/activate
+pip install Flask
+```
+
+In `hello.py`, create a home route returning a basic "Hello, World!"
+
+```python
+# ab-test-demo/hello.py
+from flask import Flask
+
+app = Flask(__name__)
+
+@app.route("/")
+def hello_world():
+  return "<p>Hello, World!</p>"
+```
+
+Afterwards, create a `/blog/<string:slug>` route with a "Like" button. Add `POST` handler to the route that prints "Liked!" when clicked and returns a confirmation.
+
+```python
+# ab-test-demo/hello.py
+from flask import Flask, request
+
+# ... app, hello_world()
+
+@app.route("/blog/<string:slug>", methods=["GET", "POST"])
+def blog(slug):
+  if request.method == "GET":
+    return f"""
+      <p>Welcome to the blog post: {slug}</p>
+      <form method="post" action="/blog/{slug}">
+          <input type="submit" value="Like" name="like"/>
+      </form>
+    """
+  elif request.method == "POST":
+    return f"<p>Thanks for liking {slug}</p>"
+```
+
+Finally run `flask --app hello run` and go to `http://127.0.0.1:5000` to see your basic app running.
+
+![Hello World!](../images/tutorials/python-ab-testing//hello.png)
+
+## Adding PostHog
+
+Next, we install and set up PostHog using the [Python SDK](/docs/libraries/python).
+
+```bash
+pip install posthog
+```
+
+Initialize PostHog to our `hello.py` file using your project API key and instance address from [your project settings](https://app.posthog.com/project/settings). In our `blog` route, use PostHog to capture a "liked post" event with a property of `slug` property. Since we don’t have user IDs, we use slugs as [our "target" user IDs](/tutorials/group-page-machine-flags).
```

Comment:
Hmm can we show an example using something else other than slug? It's a bit strange to target on slug, so may confuse people

Can we target on email or id? Its the more common case, and I think will be easier to follow for people who are new.

Reply:
Updated with UUID user_id 😄